### PR TITLE
Supervised baseline.

### PR DESCRIPTION
**Pull request by jendelel**
Initial implementation. Can't be run. Missing pretrained embeddings.

### CODE REVIEW

File: supervised/data.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+from __future__ import unicode_literals, print_function, division
+from io import open
```
Comment (oskopek) [inline supervised/data.py]: Are you sure you need this?

File: supervised/data.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+BOARD_SIZE = 25
+TEAM_IDX = {
+    'BLUE': 0,
+    'RED': 1,
+    'WHITE': 2,
+    'BLACK': 3,
+}
+SPECIAL_TEAM = 4
+SOS_TOKEN = '<SOS>'
```
Comment (oskopek) [inline supervised/data.py]: Nit, opinionated: I would prefer BOS_TOKEN = '<BOS>', but up to you.

File: supervised/data.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+BOARD_SIZE = 25
+TEAM_IDX = {
+    'BLUE': 0,
+    'RED': 1,
+    'WHITE': 2,
+    'BLACK': 3,
+}
+SPECIAL_TEAM = 4
```
Comment (oskopek) [inline supervised/data.py]: Can we rename this to something like PADDING_TEAM, or whatever, and put it into the dictionary above?

File: supervised/data.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class Vocab:
+    word_to_idx = None
+    idx_to_word = None
+
+    def __init__(self, words):
+        words_unique = list(set(words))
+        if len(words) != len(words_unique):
+            print("Warining: found word duplicates!")
```
Comment (oskopek) [inline supervised/data.py]:
```suggestion
            print("Warning: Found word duplicates!")
```

File: supervised/data.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        self.idx_to_word = [EOS_TOKEN, SOS_TOKEN] + words_unique
+        self.word_to_idx = {}
```
Comment (oskopek) [inline supervised/data.py]:
```suggestion
        self.word_to_idx = {word: i for i, word in enumerate(self.idx_to_word)}
```

Comment (oskopek) [reply supervised/data.py]: and remove the bottom for loop

File: supervised/data.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        return len(self.idx_to_word)
+
+    # Turn a Unicode string to plain ASCII, thanks to
+    # http://stackoverflow.com/a/518232/2809427
+    @staticmethod
+    def _unicodeToAscii(s):
+        return ''.join(c for c in unicodedata.normalize('NFD', s) if unicodedata.category(c) != 'Mn')
+
+    @staticmethod
+    def _normalize_word(word):
```
Comment (oskopek) [inline supervised/data.py]: Could we add a function `"""` comment here, what this method is supposed to do? As in what it normalizes?

File: supervised/data.py (hunk 1 above)
Comment (oskopek) [reply supervised/data.py]: Can we not just use the default python built-in `open(...)`

File: supervised/data.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            return [Vocab._normalize_word(line) for line in f]
+
+    @staticmethod
+    def load_from_file(fname):
+        words = Vocab._load_data(fname)
+        return Vocab(words)
+
+
+def generate_board(vocab):
+    words = random.sample(vocab.idx_to_word[2:], BOARD_SIZE)  # Sample without replacement
```
Comment (oskopek) [inline supervised/data.py]:
```suggestion
    words = random.sample(vocab.idx_to_word[2:], BOARD_SIZE)  # Sample without replacement (without EOS/BOS).
```

File: supervised/data.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    # BLUE: 0; RED: 1; WHITE: 2; BLACK: 3
+    while True:
+        team_idx = np.random.randint(low=0, high=3, size=len(words), dtype=np.int64)
+        team_idx[black_index] = TEAM_IDX['BLACK']
+        unique_idx = np.unique(team_idx)
+        assert len(unique_idx) <= len(TEAM_IDX)
+        # Make sure that all teams are represented. White does not have to be represented.
+        if len(unique_idx) == 3 or (len(unique_idx) == 2 and TEAM_IDX['WHITE'] not in unique_idx):
+            break
+    yield zip(words, team_idx)
```
Comment (oskopek) [inline supervised/data.py]:
```suggestion
    return zip(words, team_idx)
```

File: supervised/data.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def generate_board(vocab):
+    words = random.sample(vocab.idx_to_word[2:], BOARD_SIZE)  # Sample without replacement
+    # Choose a black word
+    black_index = random.randrange(0, len(words))
+
+    # Assign cards to teams:
+    # BLUE: 0; RED: 1; WHITE: 2; BLACK: 3
+    while True:
```
Comment (oskopek) [inline supervised/data.py]: As far as I can tell, this does not generate the board according to Codenames rules, does it? As in, BLUE/RED have 8/9 (this number should be a parameter I guess) cards depending on who starts.

File: supervised/data.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        unique_idx = np.unique(team_idx)
+        assert len(unique_idx) <= len(TEAM_IDX)
+        # Make sure that all teams are represented. White does not have to be represented.
+        if len(unique_idx) == 3 or (len(unique_idx) == 2 and TEAM_IDX['WHITE'] not in unique_idx):
+            break
+    yield zip(words, team_idx)
+
+
+def tensors_from_board(vocab, pairs):
+    pairs = [(vocab.word_to_idx[word], team) for (word, team) in pairs]
```
Comment (oskopek) [inline supervised/data.py]:
```suggestion
    pairs = [(vocab.word_to_idx[word], team) for word, team in pairs]
```

File: supervised/data.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        for i, word, in enumerate(self.idx_to_word):
+            self.word_to_idx[word] = i
+
+    def __len__(self):
+        return len(self.idx_to_word)
+
+    # Turn a Unicode string to plain ASCII, thanks to
+    # http://stackoverflow.com/a/518232/2809427
+    @staticmethod
+    def _unicodeToAscii(s):
```
Comment (oskopek) [inline supervised/data.py]: `_unicode_to_ascii`?

File: supervised/network.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        self.embedding = nn.Embedding(input_size, hidden_size)
+        self.gru = nn.GRU(hidden_size, hidden_size)
+
+    def forward(self, input, hidden):
+        embedded = self.embedding(input).view(1, 1, -1)
+        output = embedded
+        output, hidden = self.gru(output, hidden)
+        return output, hidden
+
+    def initHidden(self):
```
Comment (oskopek) [inline supervised/network.py]: `init_hidden`

File: supervised/network.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        output = torch.cat((embedded[0], attn_applied[0]), 1)
+        output = self.attn_combine(output).unsqueeze(0)
+
+        output = F.relu(output)
+        output, hidden = self.gru(output, hidden)
+
+        output = F.log_softmax(self.out(output[0]), dim=1)
+        return output, hidden, attn_weights
+
+    def initHidden(self):
```
Comment (oskopek) [inline supervised/network.py]: init_hidden

File: supervised/train.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+        decoder_hidden = torch.cat([encoder_hidden, clue_scaled], 1)
+        decoder_input = torch.tensor([[data.SOS_TOKEN]], device=device)
+
+        for di in range(intended_length):
+            decoder_output, decoder_hidden, decoder_attention_unused = self.decoder(decoder_input, decoder_hidden, encoder_outputs)
+            topv, topi = decoder_output.topk(1)
+            decoder_input = topi.squeeze().detach()  # detach from history as input
+
+            loss += criterion(decoder_output, intended_output[di])
+            if decoder_input.item() == data.EOS_token or not team_words[decoder_input.item()]:
```
Comment (oskopek) [inline supervised/train.py]:
```suggestion
            if decoder_input.item() == data.EOS_TOKEN or not team_words[decoder_input.item()]:
```

File: supervised/train.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+    loss = 0
+
+    for ei in range(input_length):
+        encoder_output, encoder_hidden = encoder(board_input[ei], encoder_hidden)
+        encoder_outputs[ei] = encoder_output[0, 0]
+
+    clue_logits = clue_predictor(encoder_hidden)
+
+    decoder_input = torch.tensor([[data.SOS_token]], device=device)
```
Comment (oskopek) [inline supervised/train.py]:
```suggestion
    decoder_input = torch.tensor([[data.SOS_TOKEN]], device=device)
```

File: supervised/train.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+    decoder_hidden = encoder_hidden
+
+    intended_output = []
+    for di in range(9):  # At most 9 cards of the same color.
+        decoder_output, decoder_hidden, decoder_attention_unused = decoder(decoder_input, decoder_hidden, encoder_outputs)
+        topv, topi = decoder_output.topk(1)
+        decoder_input = topi.squeeze().detach()  # detach from history as input
+        intended_output.append(decoder_input.item())
+
+        if decoder_input.item() == data.EOS_token:
```
Comment (oskopek) [inline supervised/train.py]:
```suggestion
        if decoder_input.item() == data.EOS_TOKEN:
```

File: supervised/train.py [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+    loss += player_trainer.train_step(encoder_hidden, encoder_outputs, clue_logits, intended_output_idx, criterion, team_words)
+
+    loss.backward()
+
+    encoder_opt.step()
+    decoder_opt.step()
+    player_opt.step()
+
+
+def trainIters(encoder,
```
Comment (oskopek) [inline supervised/train.py]:
```suggestion
def train_iters(encoder,
```

File: supervised/train.py [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+    data.generate_board(board_vocab)
+    criterion = nn.NLLLoss()
+
+    for iter in range(1, n_iters + 1):
+        board = data.generate_board(board_vocab)
+        training_input = data.tensors_from_board(board_vocab, board)
+
+        # Choose team
+        team = data.TEAM_IDX['BLUE'] if random.random() < 0.5 else data.TEAM_IDX['RED']
+        team_words = [team == t for (_, t) in board]
```
Comment (oskopek) [inline supervised/train.py]: `team_words_mask`?

File: supervised/util.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+def timeSince(since, percent):
+    now = time.time()
+    s = now - since
+    es = s / (percent)
+    rs = es - s
+    return '%s (- %s)' % (asMinutes(s), asMinutes(rs))
+
+
+def showPlot(points):
```
Comment (oskopek) [inline supervised/util.py]: show_plot

File: supervised/util.py [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+import matplotlib.ticker as ticker  # noqa: E402
+
+
+def asMinutes(s):
+    m = math.floor(s / 60)
+    s -= m * 60
+    return '%dm %ds' % (m, s)
+
+
+def timeSince(since, percent):
```
Comment (oskopek) [inline supervised/util.py]: time_since

File: supervised/util.py [hunk 21]
```diff
@@ -0,0 +1,29 @@
+import time
+import math
+
+import matplotlib.pyplot as plt
+plt.switch_backend('agg')
+import matplotlib.ticker as ticker  # noqa: E402
+
+
+def asMinutes(s):
```
Comment (oskopek) [inline supervised/util.py]: as_minutes

File: supervised/util.py [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    m = math.floor(s / 60)
```
Comment (oskopek) [inline supervised/util.py]:
```suggestion
    m = s // 60
```

File: supervised/util.py [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    s -= m * 60
+    return '%dm %ds' % (m, s)
+
+
+def timeSince(since, percent):
+    now = time.time()
+    s = now - since
+    es = s / (percent)
```
Comment (oskopek) [inline supervised/util.py]:
```suggestion
    es = s / percent
```

File: supervised/util.py [hunk 24]
```diff
@@ -0,0 +1,29 @@
+import time
+import math
```
Comment (oskopek) [inline supervised/util.py]:
```suggestion
```

**Comment by jendelel** [comment]
I will do it. I also found a bug there. Probably will be done with the comments by tonight.

File: supervised/data.py (hunk 3 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/data.py (hunk 1 above)
Comment (jendelel) [reply supervised/data.py]: I don't know. I guess we can always put it back...

File: supervised/data.py (hunk 2 above)
Comment (jendelel) [reply supervised/data.py]: I was consistent with the tutorial... Perhaps they use EOS in the PyTorch world...

Comment (oskopek) [reply supervised/data.py]: EOS is fine, BOS vs SOS is what I meant. Overall I don't care.

File: supervised/data.py (hunk 4 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/data.py (hunk 1 above)
Comment (oskopek) [reply supervised/data.py]: Honestly don't care, your call.

File: supervised/data.py (hunk 5 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/data.py (hunk 2 above)
Comment (jendelel) [reply supervised/data.py]: So far keeping this one.

File: supervised/data.py (hunk 11 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/data.py (hunk 6 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/data.py (hunk 7 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/data.py (hunk 9 above)
Comment (jendelel) [reply supervised/data.py]: We don't want that right? We are simulating any state during the game. That is even variable number of cards. I added the variability of initial number of cards.

File: supervised/data.py (hunk 8 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/data.py (hunk 10 above)
Comment (jendelel) [reply supervised/data.py]: Done

File: supervised/network.py (hunk 12 above)
Comment (jendelel) [reply supervised/network.py]: Done

File: supervised/network.py (hunk 13 above)
Comment (jendelel) [reply supervised/network.py]: Done

File: supervised/data.py (hunk 9 above)
Comment (oskopek) [reply supervised/data.py]: Good point.

File: supervised/train.py [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+          att_max_len=network.ATTN_MAX_LENGTH):
+    encoder_hidden = encoder.init_hidden()
+    encoder_opt.zero_grad()
+    decoder_opt.zero_grad()
+    player_opt.zero_grad()
+
+    input_length = board_input.size(0)
+    encoder_outputs = torch.zeros(att_max_len, encoder.hidden_size, device=device)
+
+    loss = 0
```
Comment (oskopek) [inline supervised/train.py]: I don't understand the loss variable in this function. Why do you define it here as 0, then add one value (?) below, and then call a function on the number?

Comment (jendelel) [reply supervised/train.py]: I add a value after every step! I don't think it's number after the first addition is probably turns into some tensor... This is directly form the seq2seq tutorial.

**Comment by jendelel** [comment]
Ok, let's merge it then!

**Comment by jendelel** [comment]
Can you merge?

**Comment by oskopek** [comment]
Can you not merge, btw?

File: supervised/train.py (hunk 25 above)
Comment (oskopek) [reply supervised/train.py]: For the tensor, ok. But where do you modify it more than once?

**Comment by jendelel** [comment]
No I can't merge! The loop is the rnn

**Comment by oskopek** [comment]
Ah sorry, fixed, you should be able to merge now. But I still don't see why the loss is correct. On the line I commented you set it to 0. Then you, !once! do += train_step, and then you call .backward on it. That's it. Somehow that doesn't sound correct :/ But go ahead and merge if you want.
